### PR TITLE
Upgrade to PHP7 strict mode

### PR DESCRIPTION
**Pull request by eddiejaoude**
# Do NOT merge (yet)

* [x] Vagrant and Ansible

### Creational

* [ ] AbstractFactory
* [ ] Builder
* [x] FactoryMethod
* [ ] Multiton
* [ ] Pool
* [x] Prototype
* [ ] SimpleFactory
* [x] Singleton
* [x] StaticFactory

### Structural

* [ ] Adapter
* [ ] Bridge
* [ ] Composite
* [ ] DataMapper
* [ ] Decorator
* [ ] DependencyInjection
* [ ] Facade
* [ ] FluentInterface
* [ ] Proxy
* [ ] Registry

### Behavioral

* [ ] ChainOfResponsibilities
* [ ] Command
* [ ] Iterator
* [ ] Mediator
* [ ] Memento
* [ ] NullObject
* [ ] Observer
* [ ] Specification
* [ ] State
* [ ] Strategy
* [ ] TemplateMethod
* [ ] Visitor

### [More](More)
* [ ] Delegation
* [ ] ServiceLocator
* [ ] Repository

### CODE REVIEW

**Comment by eddiejaoude** [comment]
How to define class variable types?

**Comment by bocharsky-bw** [comment]
Does it possible in PHP7? I think it could to do only with annotations only

**Comment by eddiejaoude** [comment]
I am not sure. I couldn't find anything on it while having a quick search. But I would have thought that if in strict mode class variables need a declared type too.

**Comment by bocharsky-bw** [comment]
In HHVM properties declare with scalar types as:
``` php
class User {
    private string $name;
}
```

**Comment by eddiejaoude** [comment]
@bocharsky-bw unfortunately that didn't work :cry: . Thanks though :smile:

File: Creational/Prototype/Tests/PrototypeTest.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        $this->barBookPrototype = new BarBookPrototype();
+    }
+
+    public function testCloneOfFooBookPrototype()
+    {
+        for ($i = 0; $i < 100; $i++) {
+            $book = clone $this->fooBookPrototype;
+
+            $this->assertInstanceOf('DesignPatterns\Creational\Prototype\FooBookPrototype', $book);
+            $this->assertNotSame($this->fooBookPrototype, $book);
```
Comment (eddiejaoude) [inline Creational/Prototype/Tests/PrototypeTest.php]: Removed example usage file and replaced with test. Is this a good test, or can it be improved?

File: Creational/FactoryMethod/Bicycle.php [hunk 2]
```diff
@@ -1,9 +1,11 @@
 <?php
+declare(strict_types = 1);
 
 namespace DesignPatterns\Creational\FactoryMethod;
 
 /**
- * Bicycle is a bicycle
+ * Class Bicycle
```
Comment (marius-rizac) [inline Creational/FactoryMethod/Bicycle.php]: This type of comments is non sens. Why do you have to write `Class Bicycle` when is clearly that you define this class 3 lines below ?

File: Creational/FactoryMethod/FactoryMethod.php [hunk 3]
```diff
@@ -1,26 +1,25 @@
 <?php
+declare(strict_types = 1);
 
 namespace DesignPatterns\Creational\FactoryMethod;
 
 /**
- * class FactoryMethod
+ * Class FactoryMethod
```
Comment (marius-rizac) [inline Creational/FactoryMethod/FactoryMethod.php]: here too and for every class the same

File: Creational/FactoryMethod/FactoryMethod.php [hunk 4]
```diff
@@ -1,26 +1,25 @@
...
+ * @package DesignPatterns\Creational\FactoryMethod
  */
 abstract class FactoryMethod
 {
-
+    /** @var int */
     const CHEAP = 1;
+
+    /** @var int */
```
Comment (marius-rizac) [inline Creational/FactoryMethod/FactoryMethod.php]: definitely `FAST` and `CHEAP` are constants not variables. I would use `@const` or no docblock at all

**Comment by HassanAlthaf** [comment]
Keep up the hard work, if you need help, feel free to inform me.

**Comment by domnikl** [comment]
I am closing here as it is already very old and I am working on PHP7 feature usage in the php7 branch. Thanks @eddiejaoude for the hard work and @bocharsky-bw, @marius-rizac and @HassanAlthaf for the feedback, I already incorporated most of it in the php-7 branch

**Pull request closed without merging**